### PR TITLE
Remote shard mapping over TCP

### PR DESCRIPTION
**Pull request by otoolep**
With this change remote mapping no longer uses HTTP, as the HTTP ports exposed by nodes on the cluster are not known cluster wide. The TCP ports exposed by the cluster service are, so this change uses that functionality. Each RemoteMapper has its own dedicated connection pool for each node, and remote mapping TCP connections are in no way coupled with query TCP connections.

It is not clear to me if this will be part of 0.9.2, so I have left the CHANGELOG untouched for now.

I tested this code by forcing all mapping to take place over the network, spun up a single node, and performed some queries. This meant shard mapping actually took place over localhost. The returned data looked good.

The structure is as follows.

- Query planner requests mappers from `ShardMapper` for 1 or more shards. This takes place on the Co-ordinator node.
- `ShardMapper` returns an object implementing the `Mapper` interface.
- If a shard is local, the local store is asked for the Mapper.
- If a shard is remote, the Mapper is actually a `RemoteMapper`.
- `RemoteMapper` connects to the target node, sending the shard ID and query to the remote node.
- The Cluster service on the remote node picks up the request, and asks its local store to create a Mapper for the requested shard.
- Cluster service streams data, served by the Mapper, back to the co-ordinating node.
- `RemoteMapper` decode results and presents them to the query engine.

### CODE REVIEW

**Comment by otoolep** [comment]
@benbjohnson @dgnorton @jwilder @pauldix

File: tsdb/mapper.go [hunk 1]
```diff
@@ -26,13 +26,13 @@ func (a mapperValues) Len() int           { return len(a) }
 func (a mapperValues) Less(i, j int) bool { return a[i].Time < a[j].Time }
 func (a mapperValues) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 
-type mapperOutput struct {
+type MapperOutput struct {
```
Comment (otoolep) [inline tsdb/mapper.go]: Simply exporting this type, no other changes.

File: tsdb/query_executor.go [hunk 2]
```diff
@@ -248,7 +248,7 @@ func (q *QueryExecutor) plan(stmt *influxql.SelectStatement, chunkSize int) (Exe
 	var executor Executor
 	if len(mappers) > 0 {
 		// All Mapper are of same type, so check first to determine correct Executor type.
-		if _, ok := mappers[0].(*RawMapper); ok {
```
Comment (otoolep) [inline tsdb/query_executor.go]: Not too happy about this. This check is now in two places -- at the query planner layer, and close to the TSBD store. Works very well, but subject to bugs in the future if one place is updated and the other is not.

File: cluster/shard_mapper.go [hunk 3]
```diff
@@ -27,56 +26,81 @@ type ShardMapper struct {
...
-		m, err = r.TSDBStore.CreateMapper(sh.ID, stmt, chunkSize)
+	if sh.OwnedBy(s.MetaStore.NodeID()) && !s.ForceRemoteMapping {
+		m, err = s.TSDBStore.CreateMapper(sh.ID, stmt, chunkSize)
 		if err != nil {
 			return nil, err
 		}
 	} else {
-		rm := NewRemoteMaper(sh.OwnerIDs[0], sh.ID, stmt, chunkSize)
-		rm.MetaStore = r.MetaStore
+		conn, err := s.dial(sh.OwnerIDs[0])
```
Comment (otoolep) [inline cluster/shard_mapper.go]: The first node in the cluster that owns the shard is chosen. This needs to be replaced with the existing balancer logic. See https://github.com/influxdb/influxdb/issues/3377

File: cluster/service.go [hunk 4]
```diff
@@ -144,6 +146,12 @@ func (s *Service) handleConn(conn net.Conn) {
 				s.Logger.Printf("process write shard error: %s", err)
 			}
 			s.writeShardResponse(conn, err)
+		case mapShardRequestMessage:
+			err := s.processMapShardRequest(conn, buf)
+			if err != nil {
+				s.Logger.Printf("process map shard error: %s", err)
+				_ = writeMapShardResponseMessage(conn, NewMapShardResponse(1, err.Error()))
```
Comment (jwilder) [inline cluster/service.go]: Error ignored.  Should at least log it.

File: cluster/shard_mapper.go (hunk 3 above)
Comment (jwilder) [inline cluster/shard_mapper.go]: I'd randomize this instead if not using the balancer yet.  Hard-coding to `0` could create a hot spot on one node.

File: cluster/service.go (hunk 4 above)
Comment (otoolep) [inline cluster/service.go]: Makes sense.

File: cluster/shard_mapper.go (hunk 3 above)
Comment (otoolep) [inline cluster/shard_mapper.go]: Sounds good -- that would be a quick change.

File: cluster/shard_mapper.go [hunk 5]
```diff
@@ -84,42 +108,49 @@ func NewRemoteMaper(nodeID, shardID uint64, stmt string, chunkSize int) *RemoteM
 }
 
 // Open connects to the remote node and starts receiving data.
-func (r *RemoteMapper) Open() error {
-	node, err := r.MetaStore.Node(r.nodeID)
+func (r *RemoteMapper) Open() (err error) {
+	defer func() {
+		if err != nil {
+			r.conn.MarkUnusable()
```
Comment (jwilder) [inline cluster/shard_mapper.go]: We had a bug in the `WriteShard` case where connections were not marked as unusable correctly.  Not sure if that will happen here or not but we needed to pass the connection as a parameter to the defer func.  See: https://github.com/influxdb/influxdb/blob/master/cluster/shard_writer.go#L46

Comment (jwilder) [inline cluster/shard_mapper.go]: Actually, why is every connection marked as unusable?  That should only be called when it shouldn't be returned to the pool.  If ever conn is unusable after use, I don't see a reason to use a pool.

File: tsdb/internal/meta.pb.go [hunk 6]
```diff
@@ -16,7 +16,7 @@ It has these top-level messages:
 */
 package internal
 
-import proto "github.com/golang/protobuf/proto"
```
Comment (jwilder) [inline tsdb/internal/meta.pb.go]: Don't think changes to this file were intended.  Also, you might be using a different protobuf plugin.  Did you setup your env according to: https://github.com/influxdb/influxdb/blob/master/CONTRIBUTING.md#generated-google-protobuf-code

File: cluster/shard_mapper.go (hunk 5 above)
Comment (otoolep) [inline cluster/shard_mapper.go]: Very good -- I see what you mean. This is not correct. 

Let me fix that, so that we mark connections as unusable only under the right circumstances.

File: tsdb/internal/meta.pb.go (hunk 6 above)
Comment (otoolep) [inline tsdb/internal/meta.pb.go]: I meant to ask about this. So the existing import is the one we want, right? I want to be sure.

I will revert this part.

Comment (otoolep) [inline tsdb/internal/meta.pb.go]: Existing, as in before my change.

Comment (jwilder) [inline tsdb/internal/meta.pb.go]: Actually, it looks like you are changing it the correct import `gogo`, but it seems like this change is unrelated to the shard mapping code.

**Comment by jwilder** [comment]
A few small things but :+1: after they're fixed.

**Comment by otoolep** [comment]
Thanks for the feedback @jwilder

**Comment by otoolep** [comment]
OK, I believe I addressed all comments, merging now

**Comment by otoolep** [comment]
Oh, I just realised I merged the protobuf changes that you pointed out. I meant to handle that in a separate PR, as you requested.
